server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java

```diff
@@ -28,56 +28,91 @@
+import com.datastax.driver.core.querybuilder.Insert;
 
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 public class CassandraMailRepositoryKeysDAO {
 
     private final CassandraAsyncExecutor executor;
+    private final PreparedStatement selectKey;
     private final PreparedStatement insertKey;
     private final PreparedStatement deleteKey;
     private final PreparedStatement listKeys;
+    private final boolean strongConsistency;
 
     @Inject
-    public CassandraMailRepositoryKeysDAO(Session session) {
+    public CassandraMailRepositoryKeysDAO(Session session, CassandraConfiguration cassandraConfiguration) {
+        this.strongConsistency = cassandraConfiguration.isMailRepositoryStrongConsistency();
         this.executor = new CassandraAsyncExecutor(session);
 
+        this.selectKey = prepareGet(session);
         this.insertKey = prepareInsert(session);
         this.deleteKey = prepareDelete(session);
         this.listKeys = prepareList(session);
     }
 
+    private PreparedStatement prepareGet(Session session) {
+        return session.prepare(select(MAIL_KEY)
+            .from(KEYS_TABLE_NAME)
+            .where(eq(REPOSITORY_NAME, bindMarker(REPOSITORY_NAME)))
+            .and(eq(MAIL_KEY, bindMarker(MAIL_KEY))));
+    }
+
     private PreparedStatement prepareList(Session session) {
         return session.prepare(select(MAIL_KEY)
             .from(KEYS_TABLE_NAME)
             .where(eq(REPOSITORY_NAME, bindMarker(REPOSITORY_NAME))));
     }
 
     private PreparedStatement prepareDelete(Session session) {
-        return session.prepare(delete()
+        Delete.Where deleteStatement = delete()
             .from(KEYS_TABLE_NAME)
-            .ifExists()
             .where(eq(REPOSITORY_NAME, bindMarker(REPOSITORY_NAME)))
-            .and(eq(MAIL_KEY, bindMarker(MAIL_KEY))));
+            .and(eq(MAIL_KEY, bindMarker(MAIL_KEY)));
+
+        if (strongConsistency) {
+            return session.prepare(deleteStatement.ifExists());
+        }
+        return session.prepare(deleteStatement);
     }
 
     private PreparedStatement prepareInsert(Session session) {
-        return session.prepare(insertInto(KEYS_TABLE_NAME)
-            .ifNotExists()
+        Insert insertStatement = insertInto(KEYS_TABLE_NAME)
             .value(REPOSITORY_NAME, bindMarker(REPOSITORY_NAME))
-            .value(MAIL_KEY, bindMarker(MAIL_KEY)));
+            .value(MAIL_KEY, bindMarker(MAIL_KEY));
+
+        if (strongConsistency) {
+            return session.prepare(insertStatement.ifNotExists());
+        }
+        return session.prepare(insertStatement);
     }
 
     public Mono<Boolean> store(MailRepositoryUrl url, MailKey key) {
+        if (strongConsistency) {
+            return storeExecute(url, key);
+        }
+        return exists(url, key)
+            .filter(FunctionalUtils.identityPredicate())
+            .map(FunctionalUtils.negate())
+            .switchIfEmpty(storeExecute(url, key));
```

Comment:
Please let's not do read - before writes. I agree having a different behaviour than when we use the strong consistency mode.

That is: always store!

Reply:
But we have a test case: `storeShouldReturnFalseWhenPreviouslyStored`
And the boolean result will change the logic of `CassandraMailRepositoryCountDAO`. 
?

Reply:
> storeShouldReturnFalseWhenPreviouslyStored

When I turn off LWTs I want `storeShouldReturnTrueWhenPreviouslyStored` (idempotent behaviour)